### DISCUSSION TITLE
Re-render chart

### DISCUSSION THREAD

**Discussion by jo47011** [mhthies/smarthomeconnect#88] [Q&A]
Hi, me again 🙈 

1st of all I must mention how impressed I am on what you already have implemented.  I'm learning more day by day and it is great to see what you have realized in such a generic/extendable approach.

Here is my question: I created a filter for a chart and want to adjust the timedelta on a ChartWidget.  It works fine if I fire a reload of the entire page afterwards, e.g. using `asyncio.create_task(ws.send_json({"reload": True}))`.  I was wondering whether there is a less invasive way of forcing a redraw of the chart itself w/o reloading the page.  Calling a `await chart.render()` or alike did not succeed.  Any hint for me?

![image](https://github.com/user-attachments/assets/352f518d-b32d-4b9b-bfa2-3933f5dedd06)

**Comment by mhthies** [comment]
Hi,
sorry for the (very) late response. Unfortunately, this is not explicitly supported by the current design of SHC: A general design principle in SHC is that only *connectable* objects are expected to update their value dynamically. Attributes and parameters that are not exposed as a *connectable* object are not expected to change during runtime, i.e. the implementation is not expected to react properly to such changes and synchronize state (e.g. with the state of the JavaScript application in the web client).

Of course, we can discuss where it is possible to make configuration attributes of SHC objects dynamic, i.e. expose them via a new *connectable* object that can be used to update them dynamically. However, I doubt that this will be easily possible for the time `interval` attribute of the `ChartWidget`, as it would require a more complicated state management of the cached timeseries data on the client side. I also don't think that it is practical for the given usecase, as it would synchronize the displayed interval across all connected web client.

Currently, there is also no designated mechanism for direct communication between multiple web widgets on the client side. In my opinion, that would require explicit configuration via the Python code when creating the widgets, that would then be propagated to the client javascript.

Instead, I would propose to extend the `ChartWidget` with an optional, configurable button group for the display interval. This could look similar to your custom-built solution. The list of buttons (with interval and label) could be passed to the `ChartWidget` constructor. The buttons would only control the local chart rendering. They would not affect the data caching behviour, i.e. the widget would always query the data for the longest of the selectable ranges but configure the chart to only show the selected time interval. However, it would not store the selected interval across page reloads or back and forth navigation in the browser.

If this would be an acceptable solution, I can create an issue for this new feature and see that I find the time to implement it. Of course, you can also go for such an implementation and create a PR. :)

**Comment by jo47011** [comment]
Hi,
thanks for your detailed reply. I understand your concerns/approach.  Actually your comment: `it would synchronize the displayed interval across all connected web client` is totally correct thus my approach is not feasible for everyone.  

I'm totally fine w/ extending the `ChartWidget`.  I would kindly ask you to file the according issue since you have a better in depth knowledge of what needs to be done.  Whoever of us will find the time can implement in the future.  Is that okay for you?

**Reply by mhthies** [reply]
I've created #98.
